File: keyValuePair - Dictionary/Java/src/Dictionary.java
import static java.lang.System.*;

public class Dictionary <TKey, TValue>{

    private KeyValuePair [] entries;
    private int initialSize;
    private int entriesCount;
    public Dictionary(){
        this.initialSize = 3;
        this.entries = createArray(this.initialSize);
        this.entriesCount = 0;
    }
    @SuppressWarnings("unchecked")
    private KeyValuePair[] createArray(int size) {
        return (KeyValuePair[]) new Dictionary<?, ?>.KeyValuePair[size];
    }

    public void ResizeOrNOt(){
        if(entriesCount < entries.length ){
            return;
        }
        int newSize = entries.length + initialSize;
        System.out.println("[resize] from " + entries.length + " to " + newSize);
        KeyValuePair[] newEntries  = createArray(newSize);

        arraycopy(entries, 0, newEntries, 0, entries.length);
        entries = newEntries;

    }
    public int Size(){
        return entriesCount;
    }
    public void Set(TKey key, TValue value){
        for (KeyValuePair entry : entries) {
            if (entry!= null && entry._Key == key) {
                entry._value = value;
                return;
            }
        }
        ResizeOrNOt();
        KeyValuePair newPair = new KeyValuePair(key, value);
        entries[entriesCount] = newPair;
        entriesCount++;
    }
    public TValue Get(TKey key){
        for (KeyValuePair entry : entries) {
            if (entry!= null && entry._Key == key) {
                return entry._value;
            }
        }
        return null;
    }
    public void Remove(TKey key){
        for (int i = 0; i < entries.length; i++) {
            if (entries[i]!= null && entries[i]._Key == key){
                entries[i] = entries[entriesCount - 1];
                entries[entriesCount - 1] = null;
                entriesCount--;
                return;
            }
        }
    }
    public void Print(){
        System.out.println("-------------");
        System.out.println("[Size] : " + Size());
        if(Size() == 0 ){
            System.out.println("Dictionary is Empty");
            return;
        }
        for (int i = 0; i < entries.length; i++) {
            if(entries[i] == null){
                System.out.println("[" + i + "]");
            }
            else {
                System.out.println("[" + i + "]" + entries[i]._Key + " : " + entries[i]._value);
            }
        }
        System.out.println("============");
    }
    private class KeyValuePair{
        private TKey _Key;
        private TValue _value;
        public TKey getKey(){
            return _Key;
        }
        public TValue getValue(){
            return _value;
        }
        public void setValue(TValue _value){
            this._value = _value;
        }
        public KeyValuePair(TKey _key, TValue _value){
            this._value = _value;
            this._Key =_key;
        }
    }
}

File: C#/KeyValuePair/Dictionary.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace KeyValuePair
{
    public class Dictionary<Tkey, Tvalue> where Tkey : class
    {
        KeyValuePair[] entries;
        int initialSize;
        int entriesCount;
        public Dictionary()
        {
            initialSize = 3;
            entries = new KeyValuePair[initialSize];
        }
        public void ResizeOrNot()
        {
            if (entriesCount < entries.Length - 1)
                return;
            int newSize = entries.Length + initialSize;
            Console.WriteLine($"Resize from {entries.Length} to {newSize}");
            KeyValuePair[] newArray = new KeyValuePair[newSize];
            Array.Copy(entries, newArray, entries.Length);
            entries = newArray;

        }
        public void Set(Tkey key, Tvalue value)
        {
            for (int i = 0; i < entries.Length; i++)
            {
                if(entries[i]!= null && key == entries[i].Key)
                {
                    entries[i].Value = value;
                    return;
                }
            }
            ResizeOrNot();
            KeyValuePair newPair = new KeyValuePair(key, value);
            entries[entriesCount++] = newPair;
        }

        public Tvalue Get(Tkey key)
        {
            for (int i = 0; i < entries.Length; i++)
            {
                if (entries[i] != null && key == entries[i].Key)
                    return entries[i].Value;
            }
            return default;
        }
        public bool Remove(Tkey key)
        {
            for (int i = 0; i < entries.Length; i++)
            {
                if (entries[i] != null && key == entries[i].Key)
                {
                    entries[i] = entries[entriesCount - 1] ;
                    entries[entriesCount - 1] = null;
                    entriesCount--;
                    return true;
                }
            }
            return false;

        }
        public void Print()
        {
            Console.WriteLine("----------");
            Console.WriteLine("[size] " + Size());
            for (int i = 0; i < entries.Length; i++)
            {
                if (entries[i] == null)
                {
                    Console.WriteLine("[" + i + "]");
                    continue;
                }
                else
                {
                    Console.WriteLine("[" + i + "]" + this.entries[i].Key
                      + ":" + this.entries[i].Value);
                }
            }
            Console.WriteLine("==========");
        }
        public int Size()
        {
            return entriesCount;
        }
        class KeyValuePair
        {
            Tkey _key;
            Tvalue _value;
            public Tkey Key
            {
                get { return _key; }
            }
            public Tvalue Value
            {
                get { return _value; }
                set { _value = value; }

            }
            public KeyValuePair(Tkey key, Tvalue val)
            {
                _key = key;
                _value = val;

            }

        }

    }
}
